Shipping Labels Purchase: unit tests

hichamboushaba opened the pull request:
This just adds unit tests of the shipping labels purchase feature, and closes #3698.

Shouldn't be merged until #3765 is merged.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

hichamboushaba committed:
Create generators for mocked data

hichamboushaba committed:
Add unit tests for the purchase flow

hafizrahman commented:
Aside: all 8 test on `EditShippingLabelPackagesViewModelTest.kt` failed for me, giving this error:

```
java.lang.NoSuchMethodError: org.wordpress.android.fluxc.model.WCProductModel.getStockQuantity()I
```

This is because I recently introduced [a change on FluxC](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1931) changing the `stockQuantity` type to be double instead of Int. The accompanying change in Woo Mobile is in #1931.

Incorporating the changes from there will make the tests run fine again, I think.

hafizrahman commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/CreateShippingLabelViewModelTest.kt:
```diff
@@ -264,4 +290,81 @@ class CreateShippingLabelViewModelTest : BaseUnitTest() {
+    }
+
+    @Test
+    fun `fulfill order after purchase`() = coroutinesTestRule.testDispatcher.runBlockingTest {
```
Small nitpick: I was confused about what this is testing. How about renaming it to something like "Fulfill order after successful purchase"?

hafizrahman commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/CreateShippingLabelViewModelTest.kt:
```diff
@@ -264,4 +290,81 @@ class CreateShippingLabelViewModelTest : BaseUnitTest() {
+    }
+
+    @Test
+    fun `fulfill failed after purchase`() = coroutinesTestRule.testDispatcher.runBlockingTest {
```
Another nitpick: maybe rename to "Inform user after failed purchase"?

hafizrahman commented:
Overall I like the amount of different situations being tested! 👍

hichamboushaba committed:
Rename tests

hichamboushaba commented:
Thanks @hafizrahman for the review.

> This is because I recently introduced a change on FluxC changing the stockQuantity type to be double instead of Int. The accompanying change in Woo Mobile is in #1931.

Normally your change shouldn't impact the tests, because the commit hash included in my branch is before yours. What happens sometimes, is that after switching branches, Android Studio keeps using the old dependencies, so you have to use either `Reload Gradle project`, or restart it to fix the issue.
For our case, I merged develop in the base branch, and rebased this one on it, so it includes your change now. But as I rebased it, you won't be able to pull changes directly if you already checked it out, you can bypass this by deleting your local branch, and rechecking it out.
I also renamed the test in: 7b478ec

Ready for another round 🙏

hafizrahman commented:
> What happens sometimes, is that after switching branches, Android Studio keeps using the old dependencies, so you have to use either Reload Gradle project, or restart it to fix the issue.
For our case, I merged develop in the base branch, and rebased this one on it, so it includes your change now. But as I rebased it, you won't be able to pull changes directly if you already checked it out, you can bypass this by deleting your local branch, and rechecking it out.

Very fascinating! Thanks for teaching me about that.

Everything LGTM now 👍  but I'm not merging yet as requested in the original post.

hichamboushaba commented:
Thanks @hafizrahman 🙏 , the PR is ready for merge now.

hafizrahman reviewed (approved):
All good.
